src/librustc_codegen_ssa/mir/rvalue.rs

```diff
@@ -192,7 +192,7 @@ impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                             }
                         }
                     }
-                    mir::CastKind::Pointer(PointerCast::ClosureFnPointer(_)) => {
+                    mir::CastKind::Pointer(PointerCast::ClosureFnPointer(_, _)) => {
```

Comment:
This is wrong, it needs a shim to work correctly. I think we should consider having a coercion from a Rust function item (*not* a `fn` pointer) to an `extern "..." fn` pointer.

Then the closure coercion would still be "like coercing `<typeof(|| ...) as FnOnce<_>>::call_once`, modulo first argument", and they would both have the extra capability of switching ABI.